bundles/tools.vitruv.dsls.reactions.runtime/src/tools/vitruv/dsls/reactions/runtime/routines/AbstractRoutine.xtend

```diff
@@ -0,0 +1,208 @@
+package tools.vitruv.dsls.reactions.runtime.routines
+
+import java.io.IOException
+import java.util.function.Function
+import org.eclipse.emf.ecore.EObject
+import org.eclipse.emf.ecore.resource.Resource
+import org.eclipse.emf.ecore.util.EcoreUtil
+import tools.vitruv.dsls.reactions.runtime.helper.PersistenceHelper
+import static extension tools.vitruv.dsls.reactions.runtime.helper.ReactionsCorrespondenceHelper.getCorrespondingElements
+import static extension tools.vitruv.dsls.reactions.runtime.helper.ReactionsCorrespondenceHelper.addCorrespondence
+import static extension tools.vitruv.dsls.reactions.runtime.helper.ReactionsCorrespondenceHelper.removeCorrespondences
+import tools.vitruv.dsls.reactions.runtime.structure.CallHierarchyHaving
+import tools.vitruv.dsls.reactions.runtime.structure.Loggable
+import tools.vitruv.change.propagation.ResourceAccess
+import org.eclipse.emf.common.util.URI
+import static com.google.common.base.Preconditions.checkState
+import tools.vitruv.dsls.reactions.runtime.state.ReactionExecutionState
+import org.eclipse.xtend.lib.annotations.Accessors
+import tools.vitruv.change.correspondence.CorrespondenceModel
+
+abstract class AbstractRoutine extends CallHierarchyHaving implements Routine {
+	val AbstractRoutinesFacade routinesFacade
+	@Accessors(PROTECTED_GETTER)
+	extension val ReactionExecutionState executionState
+
+	new(AbstractRoutinesFacade routinesFacade, ReactionExecutionState executionState, CallHierarchyHaving calledBy) {
+		super(calledBy)
+		this.routinesFacade = routinesFacade
+		this.executionState = executionState
+	}
+
+	// generic return type for convenience; the requested type has to match the type of the facade provided during construction
+	protected def <T extends AbstractRoutinesFacade> T getRoutinesFacade() {
+		return routinesFacade as T
+	}
```

Comment:
Why can't we make a generic class to have this explicit?

Reply:
This was a quite complicated thing. I will have another look it at, because I am not sure anymore whether this is conceptually necessary or just to not bloat up the code generation as we then have to add the generic type parameter basically everywhere.

Reply:
Proposal: We document that point in an issue. Then I will either address that issue in a subsequent PR or document in that issue why we cannot/should not change that, as soon as I have checked it. If this can be improved, it would further bloat up this PR and reviewing the further changes will become even harder than it was for the current changes. Do you agree, @JanWittler?

Reply:
Fine with that

Reply:
Done in #46.